ci: add continuous benchmarking

CarlWachter opened the pull request:
This will execute the benchmarks in the hermit-rs repo according to the `bench.json` file. Results will be published to Github Pages.

CI will fail until https://github.com/hermit-os/hermit-rs/pull/708 gets merged.

jounathaen commented on .github/workflows/ci.yml:
```diff
@@ -240,3 +243,54 @@ jobs:
+
+  benchmark:
+    name: Bench
+    runs-on: ubuntu-latest
```
```suggestion
    runs-on: [self-hosted]
```

jounathaen commented:
Ci is unfortunately still failing, despite https://github.com/hermit-os/hermit-rs/pull/708 being merged and also used.

CarlWachter commented:
@jounathaen it's passing now and results look good

jounathaen commented:
The benchmarks take 27 minutes now, which is 10 mins longer than the tests. Can we change the benchmark setup, so that the CI duration is not dominated by the benchmarks anymore (parallelize it on two runners or reduce the benchmark size)?

jounathaen commented:
> I think the benchmarks taking more time is fine, but I would suggest running the benchmark suite after merging and on PRs only on demand.

But in that case, we wouldn't know if a PR affects performance before merging it.

jounathaen commented:
As @mkroening pointed out, we already have gh-pages installed for this repository (https://hermit-os.github.io/kernel/hermit/). We would need to find a way of publishing the results on a different URL. Martin suggested a separate repository for gh-pages, where this action pushes to. Maybe a combination of both gh-pages into subdirectories is also possible.

mkroening commented on .github/benchmarks/netbench.json:
```diff
@@ -0,0 +1,108 @@
+    },
+    {
+        "name": "Netbench TCP BW - Client",
+        "command": "parallel ::: 'cargo run --manifest-path ./Cargo.toml --bin tcp-server-bw --release --target x86_64-unknown-linux-gnu -- --address 10.0.5.3 --bytes 1048576 --rounds 1000' 'sleep 10 && sudo qemu-system-x86_64 -display none -serial stdio -kernel hermit-loader-x86_64 -cpu qemu64,apic,fsgsbase,rdtscp,xsave,xsaveopt,fxsr,rdrand -enable-kvm -initrd target/x86_64-unknown-hermit/releas ...
```
Is there a reason for using user networking instead of taps? I get more than 15% more bandwidth on my machine.

mkroening replied:
It should work as described in the [loader's README](https://github.com/hermit-os/loader?tab=readme-ov-file#network-support). :)

mkroening commented on .github/benchmarks/netbench.json:
```diff
@@ -0,0 +1,108 @@
+    },
+    {
+        "name": "Netbench TCP Latency - Server",
+        "command": "parallel ::: 'sleep 10 && cargo run --manifest-path ./Cargo.toml --bin tcp-client-latency --release --target x86_64-unknown-linux-gnu -- --nonblocking --address 127.0.0.1 --bytes 1048576 --rounds 250' 'sudo qemu-system-x86_64 -display none -serial stdio -kernel hermit-loader-x86_64 -cpu qemu64,apic,fsgsbase,rdtscp,xsave,xsaveopt,fxsr,rdrand -enable-kvm -initrd target/x86_64-un ...
```
The forwarded ports for user networking are quite inconsistent right now. The application uses port 7878 by default and is configurable via CLI, but parts of the UDP and latency benches have 9975 hardcoded, which should be fixed.

The QEMU invocations should only forward the correct ports that are necessary.

CarlWachter replied:
I've changed the forwarded ports to be minimal now. As for the hardcoding that is in reference to the port that is being connected to, rather than the one that is binded to (which is determined by the CLI arg). Since this is explicitly a benchmark, not an example, i don't see any issue with that being hardcoded.

mkroening replied:
Ah, I see. So you mean that the remote port we connect to is inferred from CLI but the local port that we send the request from is hardcoded, right? In that case, the port should better be chosen by the operating system by specifying 0, or does Hermit not support that?

mkroening replied:
Oh, I see. I have opened an issue for supporting ephemeral ports (https://github.com/hermit-os/kernel/issues/1753). Could you add a TODO note that the hardcoded value is not relevant and should be changed to zero once possible? I am also wondering if changing 7878 to 9975 as well would be more or less confusing. 🤔

mkroening reviewed (changes requested):
Thanks for the PR! :)

I have a few requests. :D

CarlWachter commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,88 @@
+              rustup target add wasm32-wasip1
+              rustup component add rust-src --toolchain nightly-x86_64-unknown-linux-gnu
+          - name: Run benchmarks
+            uses: hermit-os/hermit-bench@kernel
```
I've added it to the actions README

CarlWachter commented on .github/workflows/publish_docs.yml:
```diff
@@ -38,5 +38,6 @@ jobs:
         with:
           target_branch: gh-pages
           build_dir: target/x86_64-unknown-none/doc
+          keep_history: true
```
Yeah you're right, I'll do that

mkroening commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,88 @@
+name: Benchmark
+
+on:
+  pull_request_target:
```
Sorry, I also missed https://github.com/hermit-os/kernel/pull/1726#issuecomment-2930066282. We should move to top-level right under `on` then, to restrict the default write access for the whole workflow, right? 

```yaml
permissions:
  pull-requests: write
```

mkroening commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,88 @@
+                echo "publish_benchmarks=false" >> $GITHUB_OUTPUT
+              fi
+          - name: Store benchmark results
+            uses: hermit-os/github-action-benchmark@hermit-rs
```
Thanks! :)

What kind of code in the action is specific to the benchmarked repo?

CarlWachter commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,88 @@
+              rustup target add wasm32-wasip1
+              rustup component add rust-src --toolchain nightly-x86_64-unknown-linux-gnu
+          - name: Run benchmarks
+            uses: hermit-os/hermit-bench@kernel
```
Changed that...

n0toose commented:
> But in that case, we wouldn't know if a PR affects performance before merging it.

Hi, just chiming in: The (relatively newer) `workflow_dispatch` event could help with triggering workflows on demand, should a reviewer think that a PR could affect performance: https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/manually-running-a-workflow

CarlWachter commented on .github/benchmarks/netbench.json:
```diff
@@ -0,0 +1,108 @@
+    },
+    {
+        "name": "Netbench TCP BW - Client",
+        "command": "parallel ::: 'cargo run --manifest-path ./Cargo.toml --bin tcp-server-bw --release --target x86_64-unknown-linux-gnu -- --address 10.0.5.3 --bytes 1048576 --rounds 1000' 'sleep 10 && sudo qemu-system-x86_64 -display none -serial stdio -kernel hermit-loader-x86_64 -cpu qemu64,apic,fsgsbase,rdtscp,xsave,xsaveopt,fxsr,rdrand -enable-kvm -initrd target/x86_64-unknown-hermit/releas ...
```
Agreed and changed

CarlWachter commented on .github/benchmarks/netbench.json:
```diff
@@ -0,0 +1,108 @@
+    },
+    {
+        "name": "Netbench TCP BW - Client",
+        "command": "parallel ::: 'cargo run --manifest-path ./Cargo.toml --bin tcp-server-bw --release --target x86_64-unknown-linux-gnu -- --address 10.0.5.3 --bytes 1048576 --rounds 1000' 'sleep 10 && sudo qemu-system-x86_64 -display none -serial stdio -kernel hermit-loader-x86_64 -cpu qemu64,apic,fsgsbase,rdtscp,xsave,xsaveopt,fxsr,rdrand -enable-kvm -initrd target/x86_64-unknown-hermit/releas ...
```
Short of switching to a different format, i couldn't think of anything. Given that these should only need to be set very rarely I think this is something we can live with

CarlWachter commented:
I've addressed most of the issues brought up now, but once again ran into issues with permissions. Namely pushing to the `hermit-bench` Repo is not permissible with the default `GITHUB_TOKEN`. 

@jounathaen  @mkroening  How would you feel about a Organisation owned PAT? It would allow the action to push to `hermit-bench` and be able to function on the standard `pull_request` trigger.

CarlWachter commented:
I've cut netbench out for now as i have not been able to test with tap devices and there is a kernel side incompatibility of the current netbench implementation with the kernel anyhow (https://github.com/hermit-os/kernel/pull/1780). 

Otherwise this PR is ready and I would recommend we merge this today without netbench.

I once again tested everything on my fork to ensure: https://github.com/CarlWachter/kernel/pull/9

CarlWachter commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,80 @@
+
+on:
+  pull_request_target:
+  merge_group:
+  push:
+    branches:
+      - main
```
I agree, since they are checked during the pull request anyways. In the unlikely case that a benchmark fails only on after the merge only one data point is lost. Changing this

CarlWachter commented on .github/benchmarks/general.yaml:
```diff
@@ -0,0 +1,71 @@
+- name: 1 core
+  bin: startup_benchmark
+  manifest_path: Cargo.toml
```
Since the benchmarks are in hermit-rs i think it makes sense to keep it relative to there. Especially since the build process is inherently relative to hermit-rs (since we are building hermit-rs with the current kernel, not the other way around) it makes sense to keep benchmark execution relative to it too. 

Or at least that seems more intuitive to me.

mkroening reviewed (approved):
> All squashed and ready to go. Any last objections?

I have pushed a commit reformatting and removing some things, that should be unnecessary (I don't think the benchmarks require extra toolchains or targets).

If you could check that this still works, you can squash that into your commit and I will merge. 🥳

CarlWachter commented:
> > All squashed and ready to go. Any last objections?
> 
> I have pushed a commit reformatting and removing some things, that should be unnecessary (I don't think the benchmarks require extra toolchains or targets).
> 
> If you could check that this still works, you can squash that into your commit and I will merge. 🥳

just ran it through and it still works and is a little faster
